File: native/core/tests/fixtures/mcp-http.mjs
// Test-only HTTP/SSE peer. The application never bundles Node.
import { createServer } from "node:http";
import { writeFileSync, appendFileSync, renameSync } from "node:fs";
const [mode, root] = process.argv.slice(2);
const sockets = new Set(),
  streams = new Set();
let deletes = 0;
const json = (res, status, value, headers = {}) => {
  res.writeHead(status, { "Content-Type": "application/json", ...headers });
  res.end(JSON.stringify(value));
};
const rpc = (res, id, result, headers = {}) =>
  json(res, 200, { jsonrpc: "2.0", id, result }, headers);
const held = (res) => {
  streams.add(res);
  res.on("close", () => streams.delete(res));
};
const tool = (name) => ({
  name,
  description: `HTTP fixture ${name}`,
  inputSchema: {
    type: "object",
    properties: { region: { type: "string", "x-mcp-header": "Region" } },
  },
});
const server = createServer(async (req, res) => {
  try {
    if (req.url === "/status")
      return json(res, 200, {
        streams: streams.size,
        sockets: sockets.size,
        deletes,
      });
    const pieces = [];
    for await (const piece of req) pieces.push(piece);
    const body = Buffer.concat(pieces).toString();
    const message = body ? JSON.parse(body) : null;
    appendFileSync(
      `${root}/requests.jsonl`,
      `${JSON.stringify({ method: req.method, path: req.url, headers: req.headers, message })}\n`,
    );
    if (req.url === "/stolen") return json(res, 500, { stolen: true });
    if (req.method === "DELETE") {
      deletes++;
      res.writeHead(204);
      res.end();
      return;
    }
    if (req.method === "GET") {
      if (mode === "legacy") {
        res.writeHead(200, { "Content-Type": "text/event-stream" });
        res.write(": connected\n\n");
        held(res);
        return;
      }
      res.writeHead(405);
      res.end();
      return;
    }
    if (
      mode === "auth" &&
      req.headers.authorization !== "Bearer http-private-fixture-key"
    ) {
      json(
        res,
        401,
        { message: "private-response-must-not-leak" },
        { "WWW-Authenticate": "Bearer private-challenge-must-not-leak" },
      );
      return;
    }
    if (mode === "redirect") {
      res.writeHead(307, { Location: "/stolen" });
      res.end();
      return;
    }
    if (message.method === "server/discover") {
      if (mode.startsWith("legacy")) {
        res.writeHead(400);
        res.end("initialize first");
        return;
      }
      if (mode === "init_hang") {
        held(res);
        return;
      }
      if (mode === "init_json_large") {
        json(res, 200, { padding: "x".repeat(1_048_577) });
        return;
      }
      if (mode === "init_malformed") {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end("private-invalid-body");
        return;
      }
      if (mode === "init_truncated") {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end('{"jsonrpc":');
        return;
      }
      rpc(res, message.id, {
        resultType: "complete",
        supportedVersions: ["2026-07-28"],
        capabilities: { tools: {} },
        ttlMs: 0,
        cacheScope: "private",
        _meta: {
          "io.modelcontextprotocol/serverInfo": {
            name: "http-fixture",
            version: "1",
          },
        },
      });
      return;
    }
    if (message.method === "initialize") {
      rpc(
        res,
        message.id,
        {
          protocolVersion: "2025-11-25",
          capabilities: { tools: {} },
          serverInfo: { name: "legacy-http-fixture", version: "1" },
        },
        { "Mcp-Session-Id": "fixture-session" },
      );
      return;
    }
    if (message.method === "tools/list") {
      if (mode === "catalog_error") {
        json(res, 200, {
          jsonrpc: "2.0",
          id: message.id,
          error: {
            code: -32603,
            message: "private-error-credential",
            data: { key: "private-error-credential" },
          },
        });
        return;
      }
      rpc(res, message.id, {
        resultType: "complete",
        tools: [
          tool("echo"),
          tool("failure"),
          tool("雪"),
          {
            name: "bad-header",
            inputSchema: {
              type: "object",
              properties: { bad: { type: "string", "x-mcp-header": "" } },
            },
          },
        ],
      });
      return;
    }
    if (message.method === "tools/call") {
      const action = message.params.arguments.action;
      if (action === "expired") {
        res.writeHead(404);
        res.end("session expired");
        return;
      }
      if (action === "redirect") {
        res.writeHead(307, { Location: "/stolen" });
        res.end();
        return;
      }
      if (action === "json_large") {
        json(res, 200, { padding: "x".repeat(1_048_577) });
        return;
      }
      if (action === "chunked_large") {
        res.writeHead(200, { "Content-Type": "application/json" });
        for (let i = 0; i < 18; i++) res.write("x".repeat(65_536));
        res.end();
        return;
      }
      if (
        [
          "hang",
          "sse_large",
          "sse_invalid",
          "sse_flood",
          "sse_incomplete",
        ].includes(action)
      ) {
        res.writeHead(200, { "Content-Type": "text/event-stream" });
        if (action === "hang") {
          res.write(": waiting\n\n");
          held(res);
          return;
        }
        if (action === "sse_large") res.end(`data: ${"x".repeat(1_048_577)}`);
        if (action === "sse_invalid")
          res.end("data: private-invalid-frame\n\n");
        if (action === "sse_flood") res.end("\n".repeat(200));
        if (action === "sse_incomplete") res.end('data: {"jsonrpc":');
        return;
      }
      const result = {
        resultType: "complete",
        content: [{ type: "text", text: "HTTP fixture 雪" }],
        structuredContent: {
          arguments: message.params.arguments,
          headers: req.headers,
        },
        isError: message.params.name === "failure",
      };
      if (action === "sse") {
        res.writeHead(200, {
          "Content-Type": "text/event-stream; charset=utf-8",
        });
        const wire = Buffer.from(
          `event: message\r\ndata: ${JSON.stringify({ jsonrpc: "2.0", id: message.id, result })}\r\n\r\n`,
        );
        // Split every UTF-8 scalar and CRLF pair across transport chunks.
        for (const byte of wire) {
          if (res.destroyed) break;
          res.write(Buffer.from([byte]));
          await new Promise((resolve) => setImmediate(resolve));
        }
        res.end();
        return;
      }
      rpc(res, message.id, result);
      return;
    }
    if (message.method?.startsWith("notifications/")) {
      res.writeHead(202);
      res.end();
      return;
    }
    json(res, 200, {
      jsonrpc: "2.0",
      id: message.id,
      error: { code: -32601, message: "Unknown method" },
    });
  } catch (error) {
    process.stderr.write(`${error}\n`);
    if (!res.headersSent) res.writeHead(500);
    res.end();
  }
});
server.on("connection", (socket) => {
  sockets.add(socket);
  socket.on("close", () => sockets.delete(socket));
});
server.listen(0, "127.0.0.1", () => {
  writeFileSync(
    `${root}/ready.pending`,
    JSON.stringify({ url: `http://127.0.0.1:${server.address().port}/mcp` }),
  );
  renameSync(`${root}/ready.pending`, `${root}/ready.json`);
});

File: native/core/tests/mcp_http.rs
#![cfg(unix)]
#[path = "fixtures/http.rs"]
mod http_peer;
use http_peer::Fixture;
use serde_json::json;
use shadowcode_core::mcp::Client;
use std::time::Duration;
use tokio_util::sync::CancellationToken;
#[tokio::test]
async fn modern_http_negotiates_metadata_and_headers_without_sessions_and_preserves_sse_utf8() {
    let f = Fixture::new("modern").await;
    let mut client = f.connect(CancellationToken::new()).await.unwrap();
    assert_eq!(client.pid(), None);
    assert_eq!(
        client.tools().len(),
        3,
        "Invalid header annotation must be excluded"
    );
    let value = client
        .call("雪", json!({"region":"東","action":"sse"}))
        .await
        .unwrap();
    assert_eq!(value["content"][0]["text"], "HTTP fixture 雪");
    assert_eq!(
        value["structuredContent"]["headers"]["mcp-name"],
        "=?base64?6Zuq?="
    );
    assert_eq!(
        value["structuredContent"]["headers"]["mcp-param-region"],
        "=?base64?5p2x?="
    );
    let failure = client.call("failure", json!({})).await.unwrap();
    assert_eq!(failure["isError"], true);
    client.close().await.unwrap();
    f.closed_streams().await;
    let requests = f.requests();
    assert_eq!(requests[0]["message"]["method"], "server/discover");
    assert!(requests.iter().all(|r| r["method"] == "POST"
        && r["message"]["method"] != "initialize"
        && r["headers"]["mcp-session-id"].is_null()));
    let call = requests
        .iter()
        .find(|r| r["message"]["method"] == "tools/call")
        .unwrap();
    assert_eq!(
        call["message"]["params"]["_meta"]["io.modelcontextprotocol/protocolVersion"],
        "2026-07-28"
    );
    assert_eq!(call["headers"]["mcp-protocol-version"], "2026-07-28");
    assert_eq!(call["headers"]["mcp-method"], "tools/call");
}
#[tokio::test]
async fn legacy_http_negotiates_session_and_deletes_it_without_repeating_expired_tool_calls() {
    for mode in ["legacy", "legacy-no-sse"] {
        let f = Fixture::new(mode).await;
        let mut client = f.connect(CancellationToken::new()).await.unwrap();
        assert_eq!(
            client.call("echo", json!({})).await.unwrap()["isError"],
            false
        );
        assert!(client
            .call("echo", json!({"action":"expired"}))
            .await
            .is_err());
        assert!(client.is_closed());
        client.close().await.unwrap();
        f.closed_streams().await;
        let requests = f.requests();
        assert_eq!(
            requests
                .iter()
                .filter(|r| r["message"]["method"] == "initialize")
                .count(),
            1
        );
        assert_eq!(
            requests
                .iter()
                .filter(|r| r["message"]["method"] == "tools/call")
                .count(),
            2
        );
        assert_eq!(
            requests.iter().filter(|r| r["method"] == "DELETE").count(),
            1
        );
        for request in requests
            .iter()
            .filter(|r| r["message"]["method"] == "tools/call")
        {
            assert_eq!(request["headers"]["mcp-session-id"], "fixture-session");
        }
    }
}
#[tokio::test]
async fn http_credentials_are_explicit_and_never_redirected_or_exposed_in_errors() {
    let mut f = Fixture::new("auth").await;
    let error = f.connect(CancellationToken::new()).await.err().unwrap();
    assert!(!format!("{error:#}").contains("private-"));
    assert_eq!(
        f.requests().len(),
        1,
        "Authentication must not trigger initialize fallback"
    );
    f.spec.bearer_token = Some("http-private-fixture-key".into());
    let mut client = f.connect(CancellationToken::new()).await.unwrap();
    assert!(client
        .call("echo", json!({"action":"redirect"}))
        .await
        .is_err());
    assert!(f.requests().iter().all(|r| r["path"] != "/stolen"));
    client.close().await.unwrap();
    let redirect = Fixture::new("redirect").await;
    assert!(redirect.connect(CancellationToken::new()).await.is_err());
    assert_eq!(redirect.requests().len(), 1);
}
#[tokio::test]
async fn http_rejects_oversized_malformed_truncated_and_flooded_peer_responses() {
    for mode in [
        "init_json_large",
        "init_malformed",
        "init_truncated",
        "catalog_error",
    ] {
        let f = Fixture::new(mode).await;
        let error = f.connect(CancellationToken::new()).await.err().expect(mode);
        assert!(!format!("{error:#}").contains("private-"));
        f.closed_streams().await;
    }
    for action in [
        "json_large",
        "chunked_large",
        "sse_large",
        "sse_invalid",
        "sse_flood",
        "sse_incomplete",
    ] {
        let f = Fixture::new("modern").await;
        let mut client = f.connect(CancellationToken::new()).await.unwrap();
        let error = client
            .call("echo", json!({"action":action}))
            .await
            .expect_err(action);
        assert!(!format!("{error:#}").contains("private-"));
        assert!(client.is_closed());
        f.closed_streams().await;
    }
}
#[tokio::test]
async fn http_initialization_and_calls_cancel_close_streams_and_never_resubmit() {
    let mut f = Fixture::new("init_hang").await;
    f.spec.timeout = Duration::from_millis(150);
    assert!(f.connect(CancellationToken::new()).await.is_err());
    f.closed_streams().await;
    let f = Fixture::new("init_hang").await;
    let spec = f.spec.clone();
    let connecting =
        tokio::spawn(async move { Client::connect_http(&spec, CancellationToken::new()).await });
    tokio::time::timeout(Duration::from_secs(1), async {
        while f.requests().is_empty() {
            tokio::time::sleep(Duration::from_millis(10)).await;
        }
    })
    .await
    .unwrap();
    connecting.abort();
    assert!(connecting.await.is_err_and(|error| error.is_cancelled()));
    f.closed_streams().await;
    for cancelled in [false, true] {
        let mut f = Fixture::new("modern").await;
        f.spec.timeout = Duration::from_millis(200);
        let cancel = CancellationToken::new();
        let mut client = f.connect(cancel.clone()).await.unwrap();
        if cancelled {
            tokio::spawn(async move {
                tokio::time::sleep(Duration::from_millis(40)).await;
                cancel.cancel();
            });
        }
        assert!(client.call("echo", json!({"action":"hang"})).await.is_err());
        client.close().await.unwrap();
        f.closed_streams().await;
        assert_eq!(
            f.requests()
                .iter()
                .filter(|r| r["message"]["method"] == "tools/call")
                .count(),
            1
        );
    }
    let f = Fixture::new("legacy").await;
    let client = f.connect(CancellationToken::new()).await.unwrap();
    drop(client);
    f.closed_streams().await;
}
#[tokio::test]
async fn invalid_http_specs_and_precancelled_connections_send_nothing() {
    let f = Fixture::new("modern").await;
    for url in [
        "file:///etc/passwd",
        "http://user:password@127.0.0.1/mcp",
        "https://example.com/mcp#fragment",
    ] {
        let mut spec = f.spec.clone();
        spec.url = url.into();
        assert!(Client::connect_http(&spec, CancellationToken::new())
            .await
            .is_err());
    }
    let mut spec = f.spec.clone();
    spec.url = "http://example.com/mcp".into();
    spec.bearer_token = Some("secret".into());
    assert!(Client::connect_http(&spec, CancellationToken::new())
        .await
        .is_err());
    spec = f.spec.clone();
    spec.bearer_token = Some("injected\r\nHeader:value".into());
    assert!(Client::connect_http(&spec, CancellationToken::new())
        .await
        .is_err());
    let cancel = CancellationToken::new();
    cancel.cancel();
    assert!(f.connect(cancel).await.is_err());
    assert!(f.requests().is_empty());
}
